psi4/driver/driver_findif.py

```diff
@@ -643,115 +734,113 @@ def recursive_gradients(n):
+                E[i, max_disp + j - 1] = displacements[key_string.format(k)]["energy"]
         # Now determine all diagonal force constants for this irrep.
-        if data["num_pts"] == 3:
-            diag_fcs = energies[:, 0] + energies[:, 1]
+        if metadict["stencil_size"] == 3:
+            diag_fcs = E[:, 0] + E[:, 1]
             diag_fcs -= 2 * ref_energy
-            diag_fcs /= (data["disp_size"]**2)
-        elif data["num_pts"] == 5:
-            diag_fcs = -energies[:, 0] + 16 * energies[:, 1] + 16 * energies[:, 2] - energies[:, 3]
+            diag_fcs /= (metadict["step"]["size"]**2)
+        elif metadict["stencil_size"] == 5:
+            diag_fcs = -E[:, 0] + 16 * E[:, 1] + 16 * E[:, 2] - E[:, 3]
             diag_fcs -= 30 * ref_energy
-            diag_fcs /= (12 * data["disp_size"]**2)
+            diag_fcs /= (12 * metadict["step"]["size"]**2)
         H_irr = np.diag(diag_fcs)
 
+        # TODO: It's a bit ugly to use the salc indices to grab the off-diagonals but the indices
+        # within the irrep to grab the diagonals. Is there a better way to do this?
+
         # Step Two: Off-diagonals
-        num_unique_off_elts = n_salcs * (n_salcs - 1) // 2
-        offdiag_energies = irrep_energies[diag_disps:]
-        # If offdiagonal elements exist, put them into groups per salc pairs
-        # ...if offdiagonal elements DON'T exist, reshaping would raise an error.
-        if offdiag_energies:
-            offdiag_energies = np.reshape(offdiag_energies, (num_unique_off_elts, -1))
-        offdiag_row_index = 0
-
-        for i, salc in enumerate(salcs):
-            for j, salc2 in enumerate(salcs[:i]):
-                offdiag_row = offdiag_energies[offdiag_row_index]
-                if data["num_pts"] == 3:
-                    fc = (+offdiag_row[0] + offdiag_row[1] + 2 * ref_energy - energies[i][0] - energies[i][1] -
-                          energies[j][0] - energies[j][1]) / (2 * data["disp_size"]**2)
-                elif data["num_pts"] == 5:
-                    fc = (-offdiag_row[0] - offdiag_row[1] + 9 * offdiag_row[2] - offdiag_row[3] - offdiag_row[4] +
-                          9 * offdiag_row[5] - offdiag_row[6] - offdiag_row[7] + energies[i][0] - 7 * energies[i][1] -
-                          7 * energies[i][2] + energies[i][3] + energies[j][0] - 7 * energies[j][1] -
-                          7 * energies[j][2] + energies[j][3] + 12 * ref_energy) / (12 * data["disp_size"]**2)
+        # We need off-diagonal energies, diagonal energies, AND the reference energy
+        # Grabbing off-diagonal energies is a pain, so once we know our SALCs...
+        # ...define offdiag_en to do that for us.
+        for i, salc in enumerate(salc_indices):
+            for j, salc2 in enumerate(salc_indices[:i]):
+                offdiag_en = lambda index: displacements["{j}: {}, {i}: {}".format(i=salc, j=salc2, *data["disps"]["off"][index])]["energy"]
```

Comment:
Should index variables be reused here (i, j)? And note that you may need to "capture" external variables if you want their values frozen into the lambda (e.g., https://github.com/MolSSI/QCElemental/blob/master/qcelemental/molparse/nucleus.py#L180) otherwise the lambda will use the values as eval-time not definition-time.

Reply:
Is this one still in the works?

Reply:
Already handled. I swapped i and j in the lambda for k and l.